src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -0,0 +1,307 @@
+ */
+public abstract class AbstractEntityAIGuard extends AbstractEntityAISkill<JobGuard>
+{
+    /**
+     * The start search distance of the guard to track/attack entities may get more depending on the level.
+     */
+    private static final double MAX_ATTACK_DISTANCE = 20.0D;
+
+    /**
+     * Distance the guard starts searching.
+     */
+    private static final int START_SEARCH_DISTANCE = 5;
+
+    /**
+     * Basic delay after operations.
+     */
+    private static final int BASE_DELAY = 1;
+
+    /**
+     * Max amount the guard can shoot arrows before restocking.
+     */
+    private static final int MAX_ATTACKS = 50;
+
+    /**
+     * Y range in which the guard detects other entities.
+     */
+    private static final double HEIGHT_DETECTION_RANGE = 10D;
+
+    /**
+     * Path that close to the patrol target.
+     */
+    private static final int PATH_CLOSE                = 3;
+
+    /**
+     * The distance the guard is searching entities in currently.
+     */
+    private int currentSearchDistance = START_SEARCH_DISTANCE;
+
+    /**
+     * The current target.
+     */
+    protected EntityLivingBase targetEntity;
+
+    /**
+     * Current goTo task.
+     */
+    private BlockPos currentPatrolTarget;
+
+    /**
+     * Containing all close entities.
+     */
+    private List<Entity> entityList;
+
+    /**
+     * Amount of arrows already shot or sword hits dealt.
+     */
+    protected int attacksExecuted = 0;
+
+    /**
+     * Sets up some important skeleton stuff for every ai.
+     *
+     * @param job the job class
+     */
+    protected AbstractEntityAIGuard(@NotNull final JobGuard job)
+    {
+        super(job);
+    }
+
+
+    /**
+     * Can be overridden in implementations.
+     * <p>
+     * Here the AI can check if the armour have to be re rendered and do it.
+     */
+    @Override
+    protected void updateRenderMetaData()
+    {
+        updateArmor();
+    }
+
+    /**
+     * Goes back to the building and tries to take armour from it when he hasn't in his inventory.
+     * @return the next state to go to.
+     */
+    protected AIState goToBuilding()
+    {
+        if(!walkToBuilding())
+        {
+            final AbstractBuildingWorker workBuilding = getOwnBuilding();
+            if(workBuilding != null)
+            {
+                final TileEntityColonyBuilding chest = workBuilding.getTileEntity();
+                for (int i = 0; i < workBuilding.getTileEntity().getSizeInventory(); i++)
+                {
+                    final ItemStack stack = chest.getStackInSlot(i);
+
+                    if (stack == null)
+                    {
+                        continue;
+                    }
+
+                    if (stack.getItem() instanceof ItemArmor && worker.getItemStackFromSlot(((ItemArmor) stack.getItem()).armorType) == null)
+                    {
+                        final int emptySlot = worker.getInventoryCitizen().getFirstEmptySlot();
```

Comment:
Your loop is doing 5 Stacks at once. With the InventoryFunctions each Stack is separate. But in this case it could be that you have two stacks but only one empty